tests/ecall_ocall/enc/enc.cpp

```diff
@@ -82,11 +82,10 @@ oe_result_t enc_parallel_execution(
 
         ++(*counter);
 
-        unsigned release_val;
-        do
-        {
-            release_val = release->load(std::memory_order_acquire);
-        } while (0 != release_val);
+        // Wait for the signal from host before continuing.
+        // (wait until release becomes non-zero)
+        while (0 == release->load(std::memory_order_acquire))
+            ;
```

Comment:
Since the failures occurred in Release mode, the optimizer was most likely skipping over the code. 
Another option is to make release_val explicitly volatile.

Reply:
It is not clear to me where the different interpretation of the issue is coming from.

The enclave threads were supposed to wait for the host to set the release value to 1.
Instead they were waiting for the release value to be zero.

The fix is to wait for the correct condition.

How will making the release_val volatile fix the issue if the enclave thread is still waiting for the incorrect condition?

Reply:
My comments were on the code before your fix. 
I now see your fix. But how did the test not Fail every time?

Reply:
Before the fix:

host:
    create threads            (host.cpp:158)
    wait for all threads to increment counter to max value   (host.cpp:166)
    signal threads to complete by setting release to 1   (host.cpp:183)

enclave threads:
    increment counter    (enc.cpp:83)
    loop while value != 0 (enc.cpp:89) (bug)
    
The test will hang if immediately after the last thread increments the counter, the host gets control and changes release to 1. Now when the last thread gets control back, it will go into a loop waiting for the release value to become zero.
This scenario does not always happen; but can occasionally happen.

Reply:
thanks @anakrish